File: hydra-java/src/main/java/hydra/Lexical.java
package hydra;

import hydra.compute.Flow;
import hydra.core.Name;
import hydra.dsl.Flows;
import hydra.graph.Graph;
import hydra.graph.Primitive;

import hydra.util.Maybe;

import static hydra.dsl.Flows.bind;
import static hydra.dsl.Flows.fail;
import static hydra.dsl.Flows.getState;


/**
 * Lexical functions (which deal with named elements and primitives within graphs).
 */
public class Lexical {
    private Lexical() {
    }

    /**
     * Get a primitive from a graph by name; the primitive is not required to exist.
     * @param g the graph to search
     * @param name the name of the primitive
     * @return an optional containing the primitive if found
     */
    public static Maybe<Primitive> lookupPrimitive(Graph g, Name name) {
        return Maybe.justNullable(g.primitives.get(name));
    }

    /**
     * Get a primitive from the current graph by name; the primitive is required to exist.
     * @param name the name of the primitive
     * @return a flow containing the primitive
     */
    public static  Flow<Graph, Primitive> requirePrimitive(Name name) {
        return bind(getState(), g -> {
            Maybe<Primitive> mprim = lookupPrimitive(g, name);
            return mprim.<Flow<Graph, Primitive>>map(Flows::pure)
                .orElseGet(() -> fail("no such primitive function: " + name.value));
        });
    }
}

File: hydra-haskell/src/main/haskell/Hydra/Staging/Lexical.hs
-- | Functions for retrieving elements and primitive functions from a graph context

module Hydra.Staging.Lexical where

import Hydra.Qnames
import Hydra.Strip
import Hydra.Core
import Hydra.Graph
import Hydra.Compute
import Hydra.Flows
import Hydra.Errors
import Hydra.Module
import Hydra.Lexical

import qualified Data.List as L
import qualified Data.Map as M
import qualified Data.Maybe as Y
import Control.Monad


dereferenceElement :: Name -> Flow Graph (Maybe Element)
dereferenceElement name = do
  g <- getState
  return $ M.lookup name (graphElements g)

fieldsOf :: Type -> [FieldType]
fieldsOf t = case stripType t of
  TypeLambda (LambdaType _ body) -> fieldsOf body
  TypeRecord rt -> rowTypeFields rt
  TypeUnion rt -> rowTypeFields rt
  _ -> []

requireElement :: Name -> Flow Graph Element
requireElement name = do
    mel <- dereferenceElement name
    case mel of
      Just el -> return el
      Nothing -> getState >>= err
  where
    err g = fail $ "no such element: " ++ unName name
        ++ ". Available elements: {" ++ L.intercalate ", " (ellipsis (unName . elementName <$> M.elems (graphElements g))) ++ "}"
      where
        showAll = False
        ellipsis = id
--        ellipsis strings = if L.length strings > 3 && not showAll
--          then L.take 3 strings ++ ["..."]
--          else strings

requirePrimitive :: Name -> Flow Graph Primitive
requirePrimitive name = do
    g <- getState
    Y.maybe err pure $ lookupPrimitive g name
  where
    err = fail $ "no such primitive function: " ++ unName name

requireTerm :: Name -> Flow Graph Term
requireTerm name = do
  mt <- resolveTerm name
  case mt of
    Nothing -> fail $ "no such element: " ++ unName name
    Just term -> pure term

-- TODO: distinguish between lambda-bound and let-bound variables
resolveTerm :: Name -> Flow Graph (Maybe Term)
resolveTerm name = do
    g <- getState
    Y.maybe (pure Nothing) recurse $ M.lookup name $ graphElements g
  where
    recurse el = case fullyStripTerm (elementData el) of
      TermVariable name' -> resolveTerm name'
      _ -> pure $ Just $ elementData el

-- Note: assuming for now that primitive functions are the same in the schema graph
schemaContext :: Graph -> Graph
schemaContext g = Y.fromMaybe g (graphSchema g)

stripAndDereferenceTerm :: Term -> Flow Graph Term
stripAndDereferenceTerm term = case fullyStripTerm term of
  TermVariable v -> do
    t <- requireTerm v
    stripAndDereferenceTerm t
  t -> pure t

toCompactName :: M.Map Namespace String -> Name -> String
toCompactName namespaces name = case mns of
    Nothing -> unName name
    Just ns -> case M.lookup ns namespaces of
      Just pre -> pre ++ ":" ++ local
      Nothing -> local
  where
    (QualifiedName mns local) = qualifyName name

withSchemaContext :: Flow Graph x -> Flow Graph x
withSchemaContext f = do
  g <- getState
  withState (schemaContext g) f
